Sources/_Helpers/URLSession+AsyncAwait.swift

```diff
@@ -0,0 +1,121 @@
+#if canImport(FoundationNetworking)
+import Foundation
+import FoundationNetworking
+
+/// A set of errors that can be returned from the
+/// polyfilled extensions on ``URLSession``
+public enum URLSessionPolyfillError: Error {
+  /// Returned when no data and no error are provided.
+  case noDataNoErrorReturned
+}
+
+/// A private helper which let's us manage the asynchronous cancellation
+/// of the returned URLSessionTasks from our polyfill implementation.
+///
+/// This is a lightly modified version of https://github.com/swift-server/async-http-client/blob/16aed40d3e30e8453e226828d59ad2e2c5fd6355/Sources/AsyncHTTPClient/AsyncAwait/HTTPClient%2Bexecute.swift#L152-L156
+/// we use this for the same reasons as listed in the linked code in that there
+/// really isn't a good way to deal with cancellation in the 'with*Continuation' functions.
+private actor URLSessionTaskCancellationHelper {
+  enum State {
+    case initialized
+    case registered(URLSessionTask)
+    case cancelled
+  }
+
+  var state: State = .initialized
+
+  init() {}
+
+  nonisolated func register(_ task: URLSessionTask) {
+    Task {
+      await actuallyRegister(task)
+    }
+  }
+
+  nonisolated func cancel() {
+    Task {
+      await actuallyCancel()
+    }
+  }
+
+  private func actuallyRegister(_ task: URLSessionTask) {
+    switch state {
+    case .registered:
+      preconditionFailure("Attempting to register another task while the current helper already has a registered task!")
+    case .cancelled:
+      // Run through any cancellation logic which should be a noop as we're already cancelled.
+      actuallyCancel()
+      // Cancel the passed in task since we're already in a cancelled state.
+      task.cancel()
+    case .initialized:
+      state = .registered(task)
+
+    }
+  }
+
+  private func actuallyCancel() {
+    state = .cancelled
+
+    switch state {
```

Comment:
note: this will now always hit the `.cancelled` case due to the location of the assignment above. the state change should probably occur after reading the existing state.

Reply:
ah yeah, my bad, fixed